### DISCUSSION TITLE
FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

### DISCUSSION THREAD

**Discussion by DouggyC** [vercel/next.js#25644] [Help]
### What version of Next.js are you using?

10.2.3

### What version of Node.js are you using?

14.15.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS big sur 11.4

### How are you deploying your application?

next dev

### Describe the Bug

```
  npm run dev
 "dev": "NODE_OPTIONS=\"--max_old_space_size=4096\" next dev"
```

```
> NODE_OPTIONS="--max_old_space_size=4096" next dev

ready - started server on 0.0.0.0:3000, url: http://localhost:3000
info  - Using webpack 5. Reason: no next.config.js https://nextjs.org/docs/messages/webpack5
event - compiled successfully
event - build page: /
wait  - compiling...
event - compiled successfully

<--- Last few GCs --->

[1804:0x1046c6000]   118760 ms: Mark-sweep (reduce) 4090.4 (4101.8) -> 4090.4 (4102.8) MB, 4605.2 / 0.0 ms  (average mu = 0.198, current mu = 0.169) allocation failure GC in old space requested


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x1012e2ba5 node::Abort() (.cold.1) [/usr/local/bin/node]
 2: 0x1000a5cb9 node::Abort() [/usr/local/bin/node]
 3: 0x1000a5e1f node::OnFatalError(char const*, char const*) [/usr/local/bin/node]
 4: 0x1001e83e7 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
 5: 0x1001e8383 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
 6: 0x100397275 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/node]
 7: 0x100398d3a v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/usr/local/bin/node]
 8: 0x100394409 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/local/bin/node]
 9: 0x100391ca1 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]
10: 0x1003926b2 v8::internal::Heap::CollectAllAvailableGarbage(v8::internal::GarbageCollectionReason) [/usr/local/bin/node]
11: 0x1003a05fe v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]
12: 0x10036df67 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/usr/local/bin/node]
13: 0x1006ed110 v8::internal::Runtime_AllocateInOldGeneration(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node]
14: 0x100a795f9 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/usr/local/bin/node]
```

### Expected Behavior

```
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
info  - Using webpack 5. Reason: no next.config.js https://nextjs.org/docs/messages/webpack5
event - compiled successfully
wait  - compiling...
event - compiled successfully
event - build page: /
wait  - compiling...
event - compiled successfully
```

but gets `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`

### To Reproduce

Error seem to exist in previous versions, I'm unsure of exact cause.

1. npx create-next-app app-name
2. add some components
3. randomly happened

**Comment by timneutkens** [comment]
Please provide the code you had when this happened, it's possible that you caused an infinite loop in React that caused the OOM.

**Reply by DouggyC** [reply]
no infinite loop, just ran `npm run dev` and says compiled successfully. then try to visit localhost:3000 and crashes

Similar issue experienced by others same versions
https://github.com/vercel/next.js/issues/25421

**Reply by timneutkens** [reply]
Please provide the actual code you're running so that we can have a look at it.

**Comment by ahmadi-akbar** [comment]
+1 same here



<--- Last few GCs --->

[16572:000002BCB3FABED0] 22479996 ms: Scavenge (reduce) 3987.3 (4026.9) -> 3986.8 (4028.1) MB, 26.1 / 0.0 ms
 (average mu = 0.372, current mu = 0.473) allocation failure
[16572:000002BCB3FABED0] 22480012 ms: Scavenge (reduce) 3987.5 (4030.1) -> 3987.0 (4030.4) MB, 7.8 / 0.0 ms
(average mu = 0.372, current mu = 0.473) allocation failure

<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF6610C3E0F napi_wrap+108911
 2: 00007FF661067E16 v8::base::CPU::has_sse+61910
 3: 00007FF661068D16 node::OnFatalError+294
 4: 00007FF661940CEE v8::Isolate::ReportExternalAllocationLimitReached+94
 5: 00007FF661925ACD v8::SharedArrayBuffer::Externalize+781
 6: 00007FF6617CF95C v8::internal::Heap::EphemeronKeyWriteBarrierFromCode+1516
 7: 00007FF6617DACEA v8::internal::Heap::ProtectUnprotectedMemoryChunks+1258
 8: 00007FF6617D7E99 v8::internal::Heap::PageFlagsAreConsistent+2457
 9: 00007FF6617CCA71 v8::internal::Heap::CollectGarbage+2033
10: 00007FF6617D3865 v8::internal::Heap::GlobalSizeOfObjects+229
11: 00007FF66180DDBB v8::internal::StackGuard::HandleInterrupts+891
12: 00007FF66151BC56 v8::internal::interpreter::JumpTableTargetOffsets::iterator::operator=+8790
13: 00007FF6619C9BAD v8::internal::SetupIsolateDelegate::SetupHeap+465229
14: 0000005603AF408E
error Command failed with exit code 134.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
